### ISSUE TITLE
MemoryLeak: Mqtt unacknowledged messages are never deleted from tracking list

### ISSUE THREAD

**Issue by BoeseB** [Azure/azure-iot-sdk-java#364]
Using the SDK is currently not possible for us, as the implementation has a memory leak where all messages are collected in a dictionary of unacknowledged messages and never removed from this dictionary again.

Afte just 12 Hours the heap of our IoT device is exhausted because of this.

Is deleting the messages in the following function the right aproach to fix this issue? 
Where to put the line? 
Delete only when listeners are present?
https://github.com/Azure/azure-iot-sdk-java/blob/cf900dcc29f7cedfba1af9d0426c1c7e13fb4d21/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/Mqtt.java#L384-L415

**Comment by timtay-microsoft** [comment]
Good catch; this is a bug on our side. I'll take a look into fixing this soon. Thanks for filing this issue.

**Comment by timtay-microsoft** [comment]
I've gone ahead and created a branch with our preliminary fix for this issue. Feel free to try it out and let me know if you are still seeing any issues.

**Comment by BoeseB** [comment]
Thanks for the quick fix. We will test it on our side.

Is it possible to send messages with no listener attached? Then i think it is a bad idea to put that line inside the if statement that checks for existing listeners.

**Comment by timtay-microsoft** [comment]
Don't worry about the listener logic in that class. That listener is strictly internal, and should always be present regardless of what you do as a user. Because of that, the line that removes messages will always be executed.

**Comment by BoeseB** [comment]
Thanks for the clarification.
We ran the application with the sdk for the last 22 hours and did not detect any further memory leaks. 
Again thank you for the fast support.
